pypeit/sciimgstack.py

```diff
@@ -0,0 +1,385 @@
+        nimg = len(files)
+        weights = np.ones(nimg)/float(nimg)
+
+        # Load
+        img_stack, ivar_stack, rn2img_stack, crmask_stack, mask_stack = self.build_stack(
+            files, self.sci_bpm, reject_cr=reject_cr)
+
+        # ToDO The bitmask is not being properly propagated here!
+        if nimg > 1:
+            img_list = [img_stack]
+            var_stack = utils.calc_ivar(ivar_stack)
+            var_list = [var_stack, rn2img_stack]
+            img_list_out, var_list_out, outmask, nused = coadd2d.weighted_combine(
+                weights, img_list, var_list, (mask_stack == 0),
+                sigma_clip=sigma_clip, sigma_clip_stack = img_stack, sigrej=sigrej, maxiters=maxiters)
+            '''
+            img = img_list_out[0]
+            ivar = utils.calc_ivar(var_list_out[0])
+            rn2img = var_list_out[1]
+            '''
+            sciImage = scienceimage.ScienceImage.from_images(self.spectrograph, self.det,
+                                                             self.par['process'], self.sci_bpm,
+                                                             img_list_out[0],
+                                                             utils.calc_ivar(var_list_out[0]),
+                                                             var_list_out[1], np.invert(outmask),
+                                                             files=files)
+            sciImage.build_mask(saturation=self.saturation, mincounts=self.mincounts)
+            '''
+            # assumes everything masked in the outmask is a CR in the individual images
+            crmask = np.invert(outmask)
+            # Create a mask for this combined image
+            #processImage = processimage.ProcessImage(None, self.spectrograph, self.det, self.proc_par)
+            #processImage.image = img
+            #processImage.rawvarframe = var_list_out[0]
+            #processImage.crmask = crmask
+            #mask = procimg.build_mask(bpm=self.sci_bpm, saturation=self.saturation, mincounts=self.mincounts)
+            mask = procimg.build_mask(self.bitmask, img, ivar, self.sci_bpm, crmask,
+                                           saturation=self.saturation, mincounts=self.mincounts)
+            '''
+        else:
+            mask = mask_stack[0, :, :]
+            crmask = crmask_stack[0, :, :]
+            img = img_stack[0, :, :]
+            ivar = ivar_stack[0, :, :]
+            rn2img = rn2img_stack[0, :, :]
+            sciImage = scienceimage.ScienceImage.from_images(self.spectrograph, self.det,
+                                                             self.par['process'], self.sci_bpm,
+                                                             img, ivar, rn2img, crmask=crmask,
+                                                             mask=mask, files=files)
+
+        return sciImage
+
+
+    def proc_diff(self, reject_cr=True, sigma_clip=False, sigrej=None, maxiters=5):
```

Comment:
Let's have a discussion about what proc_diff is doing and where this code should be. I made the mistake of writing this code into the scienceimage class because that was the place where all the necessary products existed. This code is sufficiently fundamental that it should really be a core routine. It defines the basic image processing and mask creation propagation that we do in PypeIt. In other words, it is core functionality, but its not intimately intertwined with all of our classes.  I'd like to maintain the separation of core and objects if possible for these steps, but I'm not sure that can be easily done. We should discuss the distinction between objects and core methods in this context.

Reply:
We can move the code in `ScienceImage` to core.
But also check that the CRmask is working as you intend.
There is a TODO on line 271 for you.

Reply:
The masking is not simple, it is complicated and it is written as two routines for a reason. I tried to unify them but was not able to do that, but perhaps I missed something. Kyle if you have a way to make it more compact please suggest. The functionality works very well, and I know it is correct, so we have to be careful not to break it.

Reply:
Discussed at the meeting and resolved.